### PR TITLE
i#4134 drbbdup: Avoid 2nd scratch via max_case_encoding

### PR DESCRIPTION
**Pull request by derekbruening**
Adds a new drbbdup option max_case_encoding.  When the value is small,
we can avoid a 2nd scratch register for aarchxx, since we can fit all
encodings in immediates inside compare instructions.

Adds a sanity check.
Confirmed visually in the debug logs that the 2nd scratch is gone:
Before:
```
  --------------------------------------------------
  after instrumentation:
  TAG  0x0000ffff868340c0
   +0    m4 @0x0000fffd428950e8  f900b781   str    %x1 -> +0x0168(%x28)[8byte]
   +4    m4 @0x0000fffd42894da0  d53b4200   mrs    %nzcv -> %x0
   +8    m4 @0x0000fffd42894cd8  f900af80   str    %x0 -> +0x0158(%x28)[8byte]
   +12   m4 @0x0000fffd42894c58  d28c1000   movz   $0x6080 lsl $0x00 -> %x0
   +16   m4 @0x0000fffd42894bd8  f2a85000   movk   %x0 $0x4280 lsl $0x10 -> %x0
   +20   m4 @0x0000fffd42894b10  f2dfffe0   movk   %x0 $0xffff lsl $0x20 -> %x0
   +24   m4 @0x0000fffd42894a48  f9400000   ldr    (%x0)[8byte] -> %x0
   +28   m4 @0x0000fffd42894e20  f9400000   <label>
   +28   m4 @0x0000fffd42894980  f100041f   subs   %x0 $0x0000000000000001 lsl $0x0000000000000000 -> %xzr
   +32   m4 @0x0000fffd42894900  54000001   b.ne   @0x0000fffd42894fa0[8byte]
  --------------------------------------------------
```
After:
```
  --------------------------------------------------
  after instrumentation:
  TAG  0x0000ffff9aedf0c0
   +0    m4 @0x0000fffd56f400e8  d53b4200   mrs    %nzcv -> %x0
   +4    m4 @0x0000fffd56f3fda0  f900af80   str    %x0 -> +0x0158(%x28)[8byte]
   +8    m4 @0x0000fffd56f3fcd8  d2821000   movz   $0x1080 lsl $0x00 -> %x0
   +12   m4 @0x0000fffd56f3fc58  f2aadd60   movk   %x0 $0x56eb lsl $0x10 -> %x0
   +16   m4 @0x0000fffd56f3fbd8  f2dfffe0   movk   %x0 $0xffff lsl $0x20 -> %x0
   +20   m4 @0x0000fffd56f3fb10  f9400000   ldr    (%x0)[8byte] -> %x0
   +24   m4 @0x0000fffd56f3fe20  f9400000   <label>
   +24   m4 @0x0000fffd56f3fa48  f100041f   subs   %x0 $0x0000000000000001 lsl $0x0000000000000000 -> %xzr
   +28   m4 @0x0000fffd56f3f980  54000001   b.ne   @0x0000fffd56f3ffa0[8byte]
  --------------------------------------------------
```

Issue: #4134

### CODE REVIEW

**Comment by johnfxgalea** [comment]
@derekbruening Is this ready for review?

**Comment by derekbruening** [comment]
> @derekbruening Is this ready for review?

Yes, forgot to request it.  Thanks for noticing.

File: ext/drbbdup/drbbdup.c [hunk 1]
```diff
@@ -735,14 +743,15 @@ drbbdup_encode_runtime_case(void *drcontext, drbbdup_per_thread *pt, void *tag,
...
-    drreg_is_register_dead(drcontext, DRBBDUP_SCRATCH_REG2, where,
-                           &manager->is_scratch_reg2_dead);
-    if (!manager->is_scratch_reg2_dead) {
-        /* TODO i#4134: Add a user-set flag promising that runtime case values are
-         * always <256 so we know we don't need this 2nd scratch register.
-         */
-        drbbdup_spill_register(drcontext, bb, where, DRBBDUP_SCRATCH_REG2_SLOT,
-                               DRBBDUP_SCRATCH_REG2);
+    if (opts.max_case_encoding > 0 && opts.max_case_encoding < MAX_IMMED_IN_CMP)
+        manager->is_scratch_reg2_dead = true; /* Avoid restore, since not used. */
```
Comment (johnfxgalea) [inline ext/drbbdup/drbbdup.c:747]: I personally would have introduced a new flag that denotes whether reg2 is in use rather that piggybacking on the dead flag. It makes the code more readable and maybe less error-prone  as now the dead flag can be true when it actual fact the register is not.

**Comment by johnfxgalea** [comment]
You can punt the x86_64 optimisation for `drbbdup_insert_compare_encoding_and_branch` to another PR, but it is also suitable here as it also leverages the assumption of a max encoding case (32-bit size in this case).

**Comment by derekbruening** [comment]
> You can punt the x86_64 optimisation for `drbbdup_insert_compare_encoding_and_branch` to another PR, but it is also suitable here as it also leverages the assumption of a max encoding case (32-bit size in this case).

You mean adding the use of an immediate instead of a load?  OK.  Doesn't it not depend on the max encoding: it only depends on the current case value, which never changes?

**Comment by johnfxgalea** [comment]
> > You can punt the x86_64 optimisation for `drbbdup_insert_compare_encoding_and_branch` to another PR, but it is also suitable here as it also leverages the assumption of a max encoding case (32-bit size in this case).
> 
> You mean adding the use of an immediate instead of a load? OK. Doesn't it not depend on the max encoding: it only depends on the current case value, which never changes?



> > You can punt the x86_64 optimisation for `drbbdup_insert_compare_encoding_and_branch` to another PR, but it is also suitable here as it also leverages the assumption of a max encoding case (32-bit size in this case).
> 
> You mean adding the use of an immediate instead of a load? OK. Doesn't it not depend on the max encoding: it only depends on the current case value, which never changes?

Yeah. Okay, max encoding is not needed for that optimisation; we can just check the size of each encoding.

File: ext/drbbdup/drbbdup.c (hunk 1 above)
Comment (derekbruening) [reply ext/drbbdup/drbbdup.c:747]: Done.

**Comment by johnfxgalea** [comment]
LGTM. One thing which I noticed is that `drbbdup_prepare_redirect` does not restore reg2 before redirection when handling dynamic cases. Shouldn't reg2 be restored there?

**Comment by derekbruening** [comment]
> LGTM. One thing which I noticed is that `drbbdup_prepare_redirect` does not restore reg2 before redirection when handling dynamic cases. Shouldn't reg2 be restored there?

Good catch.  Done.

Note that you didn't actually mark "Approve" (the "LGTM" string doesn't do it in this system).

**Review by johnfxgalea — APPROVED** [review]

**Pull request merged**